core/domain/learner_playlist_services_test.py

```diff
@@ -100,62 +101,76 @@ def _get_all_learner_playlist_collection_ids(self, user_id):
             learner_playlist_model.collection_ids if
             learner_playlist_model else [])
 
-    def test_mark_exploration_to_be_played_later(self):
+    def test_already_subscribed_to_exploration(self):
         self.assertEqual(
             self._get_all_learner_playlist_exp_ids(self.user_id), [])
 
-        # Add an exploration to the learner playlist.
+        # Subscribe to exploration.
+        subscription_services.subscribe_to_exploration(
+            self.user_id, 'SUBSCRIBED_ID')
+
+        # Now if we try to add the same exploration, it shouldn't be added.
         learner_progress_services.add_exp_to_learner_playlist(
-            self.user_id, self.EXP_ID_0)
+            self.user_id, 'SUBSCRIBED_ID')
+        self.assertEqual(
+            self._get_all_learner_playlist_exp_ids(
+                self.user_id), [])
+
+
+    def test_mark_exploration_to_be_played_later_with_position(self):
         self.assertEqual(
-            self._get_all_learner_playlist_exp_ids(self.user_id),
-            [self.EXP_ID_0])
+            self._get_all_learner_playlist_exp_ids(self.user_id), [])
 
-        # Add another exploration.
+        # Test adding the exploration_id if it is already in 
+	# learner_playlist.exploration_ids 
+	# Add two explorations to the learner playlist.
+        learner_progress_services.add_exp_to_learner_playlist(
+            self.user_id, self.EXP_ID_0)
         learner_progress_services.add_exp_to_learner_playlist(
             self.user_id, self.EXP_ID_1)
         self.assertEqual(
             self._get_all_learner_playlist_exp_ids(
                 self.user_id), [self.EXP_ID_0, self.EXP_ID_1])
-
-        # Add the first exploration in a different position.
+        # Add the first exploration to the second position.
         learner_progress_services.add_exp_to_learner_playlist(
             self.user_id, self.EXP_ID_0,
             position_to_be_inserted=1)
         self.assertEqual(
             self._get_all_learner_playlist_exp_ids(
-                self.user_id), [self.EXP_ID_1, self.EXP_ID_0])
+                self.user_id), [self.EXP_ID_1, self.EXP_ID_0]) 
 
-        # Add the first exploration in the first position.
-        learner_progress_services.add_exp_to_learner_playlist(
-            self.user_id, self.EXP_ID_0,
-            position_to_be_inserted=0)
+        # Clear the playlist.
```

Comment:
This test is too long and it seems to be doing a bunch of things. I suggest:

- All the stuff above this seems like it should be in one test.
- If you want to test clearing the playlist, do that in its own test.
- If you want to test limits, do that in yet another separate and clearly-defined test.

Reply:
Got it.